include/fuzzuf/algorithms/nautilus/fuzzer/queue.hpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ * 
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file queue.hpp
+ * @brief Corpus queue of Nautilus.
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+#ifndef FUZZUF_INCLUDE_ALGORITHMS_NAUTILUS_FUZZER_QUEUE_HPP
+#define FUZZUF_INCLUDE_ALGORITHMS_NAUTILUS_FUZZER_QUEUE_HPP
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+#include "fuzzuf/algorithms/nautilus/grammartec/recursion_info.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/tree.hpp"
+#include "fuzzuf/feedback/put_exit_reason_type.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::fuzzer {
+
+using namespace fuzzuf::algorithm::nautilus::grammartec;
+
+using InitState = size_t;
+using DetState = std::pair<size_t, size_t>;
+using RandomState = std::monostate;
+
+struct QueueItem {
+  QueueItem() = delete;
+  QueueItem(size_t id,
+            Tree&& tree,
+            std::unordered_set<size_t>&& fresh_bits,
+            const std::vector<uint8_t>&& all_bits,
+            PUTExitReasonType exit_reason,
+            uint64_t execution_time)
+    : id(id),
+      tree(std::move(tree)),
+      fresh_bits(std::move(fresh_bits)),
+      all_bits(std::move(all_bits)),
+      exit_reason(exit_reason),
+      state(InitState(0)),
+      recursions(std::nullopt),
+      execution_time(execution_time) {}
+
+  size_t id;
+  Tree tree;
+  std::unordered_set<size_t> fresh_bits;
+  std::vector<uint8_t> all_bits;
+  PUTExitReasonType exit_reason;
+  std::variant<InitState, DetState, RandomState> state;
+  std::optional<std::vector<RecursionInfo>> recursions;
+  uint64_t execution_time;
+};
+
+class Queue {
+public:
+  Queue(std::string work_dir) : _current_id(0), _work_dir(work_dir) {}
+  const std::vector<QueueItem>& inputs() const { return _inputs; }
+  size_t size() const { return _inputs.size(); }
+
+  void Add(Tree&& tree,
+           const std::vector<uint8_t>&& all_bits,
+           PUTExitReasonType exit_reason,
+           Context& ctx,
+           uint64_t execution_time);
+  QueueItem Pop();
+  bool IsEmpty() const;
+  void Finished(QueueItem&& item);
+  void NewRound();
+
+private:
+  std::vector<QueueItem> _inputs;
+  std::vector<QueueItem> _processed;
```

Comment:
I think `sizeof(QueueItem)` is not so small. It would be around 100 bytes, and copying it can affect the performance largely. So I recommend you to change all of these with `std::unique_ptr<QueueItem>` if possible.

Reply:
Fixed in [2d7994e](https://github.com/fuzzuf/fuzzuf/pull/23/commits/2d7994e50b3bc9487e4bd280a42d5d1efc95410c)

Reply:
As a side note, unique_ptr has `operator->` so you can use unique_ptr the same way as raw pointers.
Anyways good.

Reply:
Modified in [77dc662](https://github.com/fuzzuf/fuzzuf/pull/23/commits/77dc662bf898f43abc8ecc92ec76ae6a602186df)

Reply:
Great.